src/SSTFrequencyLimitDetector.cpp

```diff
@@ -0,0 +1,209 @@
+        , m_bucket_sse_frequency{
+            m_platform_io.read_signal(
+                    "SST::HIGHPRIORITY_FREQUENCY_SSE:0", GEOPM_DOMAIN_BOARD, 0),
+            m_platform_io.read_signal(
+                    "SST::HIGHPRIORITY_FREQUENCY_SSE:1", GEOPM_DOMAIN_BOARD, 0),
+            m_platform_io.read_signal(
+                    "SST::HIGHPRIORITY_FREQUENCY_SSE:2", GEOPM_DOMAIN_BOARD, 0)}
+        , m_bucket_avx2_frequency{
+            m_platform_io.read_signal(
+                    "SST::HIGHPRIORITY_FREQUENCY_AVX2:0", GEOPM_DOMAIN_BOARD, 0),
+            m_platform_io.read_signal(
+                    "SST::HIGHPRIORITY_FREQUENCY_AVX2:1", GEOPM_DOMAIN_BOARD, 0),
+            m_platform_io.read_signal(
+                    "SST::HIGHPRIORITY_FREQUENCY_AVX2:2", GEOPM_DOMAIN_BOARD, 0)}
+        , m_bucket_avx512_frequency{
+            m_platform_io.read_signal(
+                    "SST::HIGHPRIORITY_FREQUENCY_AVX512:0", GEOPM_DOMAIN_BOARD, 0),
+            m_platform_io.read_signal(
+                    "SST::HIGHPRIORITY_FREQUENCY_AVX512:1", GEOPM_DOMAIN_BOARD, 0),
+            m_platform_io.read_signal(
+                    "SST::HIGHPRIORITY_FREQUENCY_AVX512:2", GEOPM_DOMAIN_BOARD, 0)}
+        , m_sse_hp_tradeoffs()
+        , m_avx2_hp_tradeoffs()
+        , m_avx512_hp_tradeoffs()
+        , m_cores_in_packages()
+        , m_core_frequency_limits(m_core_count, {{m_core_count / m_package_count, m_cpu_frequency_max}})
+        , m_core_lp_frequencies(m_core_count, m_cpu_frequency_sticker)
+    {
+        for (size_t i = 0; i < m_core_count; ++i) {
+            m_clos_association_signals.push_back(m_platform_io.push_signal(
+                "SST::COREPRIORITY:ASSOCIATION", GEOPM_DOMAIN_CORE, i));
+        }
+        for (unsigned int i = 0; i < m_package_count; ++i) {
+            m_sst_tf_enable_signals.push_back(m_platform_io.push_signal(
+                "SST::TURBO_ENABLE:ENABLE", GEOPM_DOMAIN_PACKAGE, i));
+            const auto nested_ctls = platform_topo.domain_nested(
+                GEOPM_DOMAIN_CORE, GEOPM_DOMAIN_PACKAGE, i);
+            m_cores_in_packages.push_back(
+                std::vector<int>(nested_ctls.begin(), nested_ctls.end()));
+        }
+
+        for (size_t bucket = 0; bucket < m_bucket_hp_cores.size(); ++bucket) {
+            m_sse_hp_tradeoffs.emplace_back(
+                m_bucket_hp_cores[bucket], m_bucket_sse_frequency[bucket]);
+            m_avx2_hp_tradeoffs.emplace_back(
+                m_bucket_hp_cores[bucket], m_bucket_avx2_frequency[bucket]);
+            m_avx512_hp_tradeoffs.emplace_back(
+                m_bucket_hp_cores[bucket], m_bucket_avx512_frequency[bucket]);
+        }
+    }
+
+    void SSTFrequencyLimitDetector::update_max_frequency_estimates(
+        const std::vector<double> &observed_core_frequencies)
+    {
+        for (size_t package_idx = 0; package_idx < m_package_count; ++package_idx) {
```

Comment:
Need to evaluate the direction of this search: Optimistic vs pessimistic peak freq estimation.

Reply:
Does not impact the agent, which avoids depending on scenarios where other throttling points are the main factor. However, the interface allows other users of this class to interact in that way. Need to add the following:

If the core is limited by a given cap, start at the highest license level whose max freq is <= the cap (instead of always starting at avx512)